x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java

```diff
@@ -264,7 +282,15 @@ public SegmentsStats segmentsStats(boolean includeSegmentFileSizes, boolean incl
         } else {
             return super.segmentsStats(includeSegmentFileSizes, includeUnloadedSegments);
         }
+    }
 
+    @Override
+    protected void closeNoLock(String reason, CountDownLatch closedLatch) {
+        super.closeNoLock(reason, closedLatch);
+        synchronized(closedListeners) {
```

Comment:
why are we synchonizing here?

Reply:
As this is `closeNoLock`, i.e., it can be called multiple times (and in parallel).